### PR TITLE
Multilingual compatibility and batching for Code Evaluation

### PR DESCRIPTION
**Pull request by rishab-partha**
# What does this PR do?

Adds support for C, C++ and JS and batching to Code Evaluation

# What issue(s) does this change relate to?

Code Evaluation Deployment

### CODE REVIEW

File: composer/utils/eval_client/eval_client.py [hunk 1]
```diff
@@ -7,28 +7,29 @@
...
                 'input': <test input>,
                 'output': <test output>,
                 'entry_point': <entry point>,
+                'language': <language>,
 
             }
 
         Args:
-            payload: the materials of the HTTPS request to the client.
+            payload: the materials of the batched HTTPS request to the client organized by prompt, beam generation, and test case.
```
Comment (mvpatel2000) [inline composer/utils/eval_client/eval_client.py:32]: Can you update docs here a bit? I don't get why its a triple nested list / it looks like the example input is now out of date

**Review by mvpatel2000 — COMMENTED** [review]
Can you post a manual test? and update daily?

**Comment by rishab-partha** [comment]
Manual test 
Screenshot 2023-08-11 at 2 57 11 PM

**Comment by rishab-partha** [comment]
Sample datasets are in my PR to foundry

File: composer/utils/eval_client/eval_client.py (hunk 1 above)
Comment (rishab-partha) [reply composer/utils/eval_client/eval_client.py]: fixed

File: composer/utils/eval_client/eval_client.py [hunk 2]
```diff
@@ -7,28 +7,31 @@
...
 class EvalClient(abc.ABC):
     """Abstract class for implementing eval clients, such as LambdaEvalClient."""
 
-    def invoke(self, payload: Dict[str, str]) -> bool:
-        """Invoke a provided dictionary payload to the client.
+    def invoke(self, payload: List[List[List[Dict[str, str]]]]) -> List[List[List[bool]]]:
+        """Invoke a provided batch of dictionary payload to the client.
 
-        For code generation, the payload is a JSON with the following attributes:
+        For code generation, the payload is a list of list of lists of JSONs with the following attributes:
```
Comment (mvpatel2000) [inline composer/utils/eval_client/eval_client.py:21]: The nesting structure should be described here. Why does it accept a list of list of lists? Please detail:
1. the breakdown (prompt group, test case, etc)
2. specify that the eval client is agnostic to this and should just iterate over each payload maintaining the parent list structure

File: composer/utils/eval_client/local_eval_client.py [hunk 3]
```diff
@@ -42,27 +51,87 @@ def update_offline_helper(self, code_gen: str, test_input: str, test_output: str
...
... (a.size()!=b.size()) return false;\n    for (int i=0;i<a.size();i++)\n    {\n        if (a[i]!=b[i]) return false;\n    }\n    return true;\n}\nbool issame(string a, string b){\n    return a == b;\n}\nbool issame(vector<float> a,vector<float>b){\n    if (a.size()!=b.size()) return false;\n    for (int i=0;i<a.size();i++)\n    {\n        if (abs(a[i]-b[i])>1e-4) return false;\n    }\n    return true;\n}\nbool issame(double a, double b){\n    return abs(a - b) < 1e-3;\n}\nbool issame(vector<vector<int>> a,vector<vector<int>> b){\n    if (a.size()!=b.size()) return false;\n\n    for (int i=0;i<a.size();i++)\n    {\n        if (a[i].size()!=b[i].size()) return false;\n        for (int j=0;j<a[i].size();j++)\n            if (a[i][j]!=b[i][j]) return false;\n    }\n    return true;\n}\nbool issame(map<char,int> a,map<char,int> b){\n    if (a.size()!=b.size()) return false;\n    map <char,int>::iterator it;\n    for (it=a.begin();it!=a.end();it++)\n    {\n        char w1=it->first;\n        int w2=it->second;\n        if (b.find(w1)==b.end()) return false;\n        if (b[w1]!=w2) return false;\n    }\n\n    return true;\n}\nbool issame(long long a, long long b){\n    return a == b;\n}\n'
```
Comment (mvpatel2000) [inline composer/utils/eval_client/local_eval_client.py:80]: Please include a comment detailing what this is

Is there a way to make it more readable...? like textwrap deedent or something?

File: composer/utils/eval_client/local_eval_client.py [hunk 4]
```diff
@@ -42,27 +51,87 @@ def update_offline_helper(self, code_gen: str, test_input: str, test_output: str
...
+            code_gen = prefix + code_gen
+
+            ending = '\nint main() {\n    cout << issame(' + entry_point + '(' + test_input + '), ' + test_output + ');\n    return 0;\n}'
```
Comment (mvpatel2000) [inline composer/utils/eval_client/local_eval_client.py:83]: same as above

File: composer/utils/eval_client/local_eval_client.py [hunk 5]
```diff
@@ -42,27 +51,87 @@ def update_offline_helper(self, code_gen: str, test_input: str, test_output: str
...
+                f.write(code_gen)
+
+            run_process = subprocess.run(['node', 'test_code.js'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            output = run_process.stdout.decode()
+            if 'test_code.js' in os.listdir():
+                os.remove('test_code.js')
+            val.value = output == 'true\n'
+
+        elif language == 'c':
+            prefix = '#include <stdio.h>\n#include <stdbool.h>\n#include <math.h>\n#include <stdlib.h>\nbool issame_int(int a, int b){\n    return a == b;\n}\nbool issame_bool(bool a, bool b){\n    return a == b;\n}\nbool issame_float(float a, float b){\n    return fabs(a-b) < 1e-4;\n}\n#define issame(a, b) _Generic((a), int: issame_int, bool: issame_bool, float: issame_float)(a, b)\n'
```
Comment (mvpatel2000) [inline composer/utils/eval_client/local_eval_client.py:115]: same as above

**Review by mvpatel2000 — COMMENTED** [review]
Some minor docs comments. Mostly LGTM

File: composer/utils/eval_client/eval_client.py (hunk 2 above)
Comment (rishab-partha) [reply composer/utils/eval_client/eval_client.py:21]: done

File: composer/utils/eval_client/local_eval_client.py (hunk 3 above)
Comment (rishab-partha) [reply composer/utils/eval_client/local_eval_client.py:80]: done

File: composer/utils/eval_client/local_eval_client.py (hunk 4 above)
Comment (rishab-partha) [reply composer/utils/eval_client/local_eval_client.py]: done

File: composer/utils/eval_client/eval_client.py [hunk 6]
```diff
@@ -7,28 +7,36 @@
...
+        """Invoke a provided batch of dictionary payload to the client.
 
-        For code generation, the payload is a JSON with the following attributes:
-            {
-                'code': <code to be evaluated>,
-                'input': <test input>,
-                'output': <test output>,
-                'entry_point': <entry point>,
+        For code generation, the payload is a list of list of lists of JSONs. The lists are organized in a nested structure, with the outer list
+        being grouped by the prompt. Then, in the nested list for each prompt, the lists are grouped by the generation beam, or generation. In the
```
Comment (mvpatel2000) [inline composer/utils/eval_client/eval_client.py:22]:
```suggestion
        being grouped by the prompt. Then, in the nested list for each prompt, the lists are grouped by the generation beam or generation. In the
```
nit: grammar

Comment (mvpatel2000) [reply composer/utils/eval_client/eval_client.py:22]: also is this complete? I don't get what generation means?

**Review by mvpatel2000 — APPROVED** [review]

File: composer/utils/eval_client/eval_client.py (hunk 6 above)
Comment (rishab-partha) [reply composer/utils/eval_client/eval_client.py]: fixed

**Pull request merged**